py_hamt/zarr_encryption_transformers.py

```diff
@@ -0,0 +1,46 @@
+from typing import Callable
+from pathlib import Path
+
+from Crypto.Cipher import ChaCha20_Poly1305
+
+# Metadata files to avoid encrypting (and decrypting)
+encryption_exclude_files = [
+    # top level meta data
+    ".zattrs",
+    ".zgroup",
+    ".zmetadata",
+    # found within variables, this includes .zattrs
+    ".zarray",
+    # important for coordinate variables, so that we can read bounds
+    "0",
+]
+
+type TransformerFN = Callable[[str, bytes], bytes]
+
+
+def create_zarr_encryption_transformers(
+    encryption_key: bytes,
+    nonce: bytes,
+) -> tuple[TransformerFN, TransformerFN]:
+    """
+    Uses XChaCha20_Poly1305 from the pycryptodome library to perform encryption, while ignoring zarr metadata files. Note that nonce must *always* be 24 bytes long.
+
+    See documentation on the encryption used at https://pycryptodome.readthedocs.io/en/latest/src/cipher/chacha20_poly1305.html
+    """
+
+    # We cannot reuse the Cipher variable for both encrypt and decrypt and must instantiate every time, since after an encryption or decryption, the cipher will return a different set of bytes for the same key and nonce.
```

Comment:
Not sure about this.

This leads me to previous comment from Phil where the nonce is the same? I would bet that nonce should be different in order to have different cipher.

Here is [the code](https://github.com/Arbol-Project/gridded-etl-tools/blob/main/gridded_etl_tools/utils/encryption.py#L110-L131) where it was done in gridded-etl-tools that could be useful, where it generates nonce every time the encryption is done, and uses the same nonce for the decryption.